### PR TITLE
Add VERIFY0P() and ASSERT0P() macros

### PR DESCRIPTION
**Pull request by dag-erling**
### Motivation and Context

As pointed out in #15224, `ASSERT0()` cannot be safely used to check a pointer.

### Description

This pull request adds `VERIFY0P()` and `ASSERT0P()` macros which can safely be used to verify that a pointer is null. It also addresses some code quality issues in the existing `VERIFY*()` macros.

### How Has This Been Tested?

The patch has been tested in a FreeBSD 15 tree.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: include/os/linux/spl/sys/debug.h [hunk 1]
```diff
@@ -125,19 +125,27 @@ spl_assert(const char *buf, const char *file, const char *func, int line)
 		if (unlikely(!(_verify3_left OP _verify3_right)))	\
 		    spl_panic(__FILE__, __FUNCTION__, __LINE__,		\
 		    "VERIFY3(" #LEFT " "  #OP " "  #RIGHT ") "		\
-		    "failed (%px " #OP " %px)\n",			\
-		    (void *) (_verify3_left),				\
-		    (void *) (_verify3_right));				\
+		    "failed (%p " #OP " %p)\n",				\
```
Comment (amotin) [inline include/os/linux/spl/sys/debug.h:128]: Unlike FreeBSD, where it was rightfully removed above, Linux kernel has special meaning for %px, that I suppose was used here.

File: include/os/linux/spl/sys/debug.h [hunk 2]
```diff
@@ -125,19 +125,27 @@ spl_assert(const char *buf, const char *file, const char *func, int line)
...
-		    "VERIFY0(0 == " #RIGHT ") "				\
+		    "VERIFY0(" #RIGHT ") "				\
 		    "failed (0 == %lld)\n",				\
-		    (long long) (_verify3_right));			\
+		    (long long)_verify0_right);				\
+	} while (0)
+
+#define	VERIFY0P(RIGHT)	do {						\
+		const uintptr_t _verify0_right = (uintptr_t)(RIGHT);	\
+		if (unlikely(!(0 == _verify0_right)))			\
```
Comment (amotin) [inline include/os/linux/spl/sys/debug.h:144]: Why not `0 != _verify0_right`?

File: lib/libspl/include/assert.h [hunk 3]
```diff
@@ -97,8 +97,8 @@ do {									\
 	const uint64_t __right = (uint64_t)(RIGHT);			\
 	if (!(__left OP __right))					\
 		libspl_assertf(__FILE__, __FUNCTION__, __LINE__,	\
-		    "%s %s %s (0x%llx %s 0x%llx)", #LEFT, #OP, #RIGHT,	\
-		    (u_longlong_t)__left, #OP, (u_longlong_t)__right);	\
+		    "%s %s %s (%p %s %p)", #LEFT, #OP, #RIGHT,		\
+		    (void *)__left, #OP, (void *)__right);		\
```
Comment (amotin) [inline lib/libspl/include/assert.h:101]: Why %p here if this is an unsigned integer assertion?

File: lib/libspl/include/assert.h [hunk 4]
```diff
@@ -107,17 +107,26 @@ do {									\
 	const uintptr_t __right = (uintptr_t)(RIGHT);			\
 	if (!(__left OP __right))					\
 		libspl_assertf(__FILE__, __FUNCTION__, __LINE__,	\
-		    "%s %s %s (0x%llx %s 0x%llx)", #LEFT, #OP, #RIGHT,	\
-		    (u_longlong_t)__left, #OP, (u_longlong_t)__right);	\
+		    "%s %s %s (%px %s %px)", #LEFT, #OP, #RIGHT,	\
+		    (void *)__left, #OP, (void *)__right);		\
```
Comment (amotin) [inline lib/libspl/include/assert.h:111]: Here %p makes sense, but why %px, considering it is user-space, where I am not sure %px is a thing on any platform.

File: include/os/linux/spl/sys/debug.h (hunk 2 above)
Comment (dag-erling) [reply include/os/linux/spl/sys/debug.h:144]: This follows the structure of the other macros. You're asserting what's inside the parentheses.

File: lib/libspl/include/assert.h (hunk 4 above)
Comment (dag-erling) [reply lib/libspl/include/assert.h:111]: Patch got mangled in transfer between trees, will regenerate.

File: include/os/freebsd/spl/sys/debug.h [hunk 5]
```diff
@@ -121,19 +123,27 @@ spl_assert(const char *buf, const char *file, const char *func, int line)
...
-		    "VERIFY0(0 == " #RIGHT ") "				\
+		    "VERIFY0(" #RIGHT ") "				\
 		    "failed (0 == %lld)\n",				\
-		    (long long) (_verify3_right));			\
+		    (long long)_verify0_right);				\
+	} while (0)
+
+#define	VERIFY0P(RIGHT)	do {						\
+		const uintptr_t _verify0_right = (uintptr_t)(RIGHT);	\
+		if (unlikely(!(0 == _verify0_right)))			\
```
Comment (amotin) [inline include/os/freebsd/spl/sys/debug.h:142]: Here could better be `NULL == ...`.

File: lib/libspl/include/assert.h [hunk 6]
```diff
@@ -97,7 +97,7 @@ do {									\
 	const uint64_t __right = (uint64_t)(RIGHT);			\
 	if (!(__left OP __right))					\
 		libspl_assertf(__FILE__, __FUNCTION__, __LINE__,	\
-		    "%s %s %s (0x%llx %s 0x%llx)", #LEFT, #OP, #RIGHT,	\
+		    "%s %s %s (%llx %s %llx)", #LEFT, #OP, #RIGHT,	\
```
Comment (amotin) [inline lib/libspl/include/assert.h:100]: Why did you remove "0x" here, but left it in above cases?

File: lib/libspl/include/assert.h [hunk 7]
```diff
@@ -107,17 +107,26 @@ do {									\
...
+		    "%s == 0 (%p == 0)", #LEFT,				\
+		    (void *)__left);					\
+} while (0)
+
+#define	VERIFY0P(LEFT)							\
+do {									\
+	const uintptr_t __left = (uintptr_t)(LEFT);			\
+	if (!(__left == 0))						\
+		libspl_assertf(__FILE__, __FUNCTION__, __LINE__,	\
+		    "%s == 0 (%p == 0)", #LEFT,				\
```
Comment (amotin) [inline lib/libspl/include/assert.h:128]: Should be NULLs also.

File: include/os/linux/spl/sys/debug.h [hunk 8]
```diff
@@ -126,18 +128,26 @@ spl_assert(const char *buf, const char *file, const char *func, int line)
...
```
Comment (amotin) [inline include/os/linux/spl/sys/debug.h:146]: "NULL == ..."

File: include/os/freebsd/spl/sys/debug.h (hunk 5 above)
Comment (dag-erling) [reply include/os/freebsd/spl/sys/debug.h:142]: No, the right-hand side is an integer, not a pointer.

File: lib/libspl/include/assert.h (hunk 6 above)
Comment (dag-erling) [reply lib/libspl/include/assert.h:100]: Unintentional, will revert

File: include/os/linux/spl/sys/debug.h (hunk 8 above)
Comment (dag-erling) [reply include/os/linux/spl/sys/debug.h:146]: No, it's an integer, not a pointer.

File: include/os/freebsd/spl/sys/debug.h (hunk 5 above)
Comment (amotin) [reply include/os/freebsd/spl/sys/debug.h:142]: Ah, it is uintptr_t, OK. But just curios why.

**Review by amotin — APPROVED** [review]

File: include/os/freebsd/spl/sys/debug.h (hunk 5 above)
Comment (dag-erling) [reply include/os/freebsd/spl/sys/debug.h:142]: There is no other type (not even `const void *`) that you can safely cast any pointer to.

**Review by oromenahar — APPROVED** [review]
check out this as well: #15224 `module/zfs/dbuf.c` both solution would work, I would prefer the other solution. But nice work.

**Comment by dag-erling** [comment]
@behlendorf could you take a look at this? a better solution for the problem fixed by #15224 plus some additional cleanup and bugfixes for the `ASSERT` and `VERIFY` macros.

**Review by behlendorf — APPROVED** [review]
Looks good to me.  Thanks for the cleanup and improvement!
